### ISSUE TITLE
How to make mlflow tracking useable on shared file systems?

### ISSUE THREAD

**Issue by Hoeze** [mlflow/mlflow#4034]
Hello, first time user here.
I am failing to use the mlflow ui because it is just awfully slow for even very low numbers of runs:
`[CRITICAL] WORKER TIMEOUT`

What I tried:
1) Changed file store to shared file system for parallel training runs (using `mlflow.set_tracking_uri()`), but recognized that mlflow by default is awfully slow this way.
Simply running `mlflow.search_runs()` for 120 runs and 9 metrics takes 30s.
2) Tried to change to sqlite URI on shared filesystem, but this causes artifacts to land in the local folder again (why??).
3) Try to find `mlflow.set_artifact_uri()` but cannot find one.

**TL;DR:**
How do I make mlflow work with a shared filesystem?
How can I store the artifacts in the same folder as the `mlflow.db` file from sqlite?

**Comment by dmatrix** [comment]
@Hoeze 

By "shared filesystem," do you mean an NFS mounted filesystem, like `/mnt/shared_directory/`?

> Changed file store to the shared file system for parallel training runs (using mlflow.set_tracking_uri()), but recognized that mlflow by default is awfully slow this way.

First, it's good to understand how the MLflow client interacts with the 
[MLflow tracking server and storage backends](https://github.com/dmatrix/mlflow-tests/tree/master/py/mlflow/server). Second, by default, if you are running everything on localhost (or your laptop), the file store is `./mlruns` directory for (MLflow entities) and the artifact repository too.

If you have hundreds of unique named experiments with hundreds of runs, and if each run has dozen or more parameters and metrics, each entity is stored under a unique run_id directory, and within that run_id directory, you will have filenames for parameters, metrics, tags, etc. So the UI is naturally is going to be slow over time, because with many directories and files, on a UNIX filesystem, this degrades.

MLflow on localhost is good for quick experimentation, but if you want to scale, use an SQLAlchemy compatible database. This would be [scenario 4](https://github.com/dmatrix/mlflow-tests/tree/master/py/mlflow/server#scenario-4)

> Tried to change to sqlite URI on shared filesystem, but this causes artifacts to land in the local folder again (why??).

Yes, as explained in [scenario 2](https://github.com/dmatrix/mlflow-tests/tree/master/py/mlflow/server#scenario-2), artifacts are stored locally under `./mlruns` or whatever `file://shared_path/ you used for artifact store, while all the MLflow entities will be stored in an SQLite database file `mlruns.db`. 

Artifact store is separate from backend entities store; only under [scenario 1 ](https://github.com/dmatrix/mlflow-tests/tree/master/py/mlflow/server#scenario-2)are they share the same filesystem directory.

> Try to find mlflow.set_artifact_uri() but cannot find one.

How do you mean?

> How do I make mlflow work with a shared filesystem?

If your shared filesystem is a remote S3 bucket or supported storage, use scenario 3.

> How can I store the artifacts in the same folder as the mlflow.db file from sqlite?

As I said, the `SQLite://<file_path>/my_sqlite.db` is an SQLite DB file where only MLflow entities (runs, parameters, tags, metadata, etc) are stored, while MLflow artifacts (files, models, etc) are stored in the artifact root directory you specify to the mlflow server as an arugment.

By default on a local host without any arguments, all go under the `./mlruns` directory from where you run your file.py (with mlflow client code in it).

I hope that answers your questions.

**Comment by hughperkins** [comment]
Just in case it's useful (I'm just another user), `[CRITICAL] WORKER TIMEOUT` => I was getting this at the start, when my mlruns was on a shared nfs file system. I changed to use postgres on a local file system, just on the server, and then now my mlflow ui just zips along smoothly :)  The artifacts are stored by the client, not by the server, so a shared file system might work well for them. I'm pretty sure the ui is slow because of the tracking db ,ie mlruns, not because of the artifacts.

**Comment by Hoeze** [comment]
Thanks for your answers :)
This issue somehow slipped my attention, sorry for that...

So what I wanted to do with mlflow was basically serverless offline tracking with SQLite3 DB:
```python
project_dir="/proj/myfancymodel"

# sqlite3:///proj/myfancymodel/mlflow.db
mlflow.set_tracking_uri("sqlite3://" + project_dir + "/mlflow.db") 
# file:///proj/myfancymodel/artifacts/
mlflow.set_artifact_uri("file://" + project_dir + "/artifacts/")
```

- sqlite3 should be more than fast enough to handle < 50,000 rows
- artifacts + database are stored at the same location
- no fancy database connection and user authentication setup
- simple file access permission is enough

This failed because there is no possibility to specify the `artifact_uri` inside a python script.

**Comment by trams** [comment]
I am not an expert but it sounds like your problem can be solved by running the server locally

> mlflow server --backend-store-uri DB_URI --artifact-store PATH_TO_THE_STORE

**Comment by epizut** [comment]
I am interested in the same feature, a serverless client that would use an SQLite3 database hosted on a shared drive, plus a custom artifact path, probably sitting in the same directory. It would allow multiple clients to generate runs from different machines, only relying on existing network drive permissions without bothering with a dedicated SQL server.
